napari/utils/events/event.py

```
@@ -501,41 +501,82 @@ def disconnect(self, callback: Union[Callback, CallbackRef, None] = None):
             self._callbacks = []
             self._callback_refs = []
         else:
-            callback = self._normalize_cb(callback)
+            callback, _pass_event = self._normalize_cb(callback)
             if callback in self._callbacks:
                 idx = self._callbacks.index(callback)
                 self._callbacks.pop(idx)
                 self._callback_refs.pop(idx)
+                self._callback_pass_env.pop(idx)
+
+    @staticmethod
+    def _get_proper_name(callback):
+        assert inspect.ismethod(callback)
+        obj = callback.__self__
+        if (
+            not hasattr(obj, callback.__name__)
+            or getattr(obj, callback.__name__) != callback
+        ):
+            # some decorators will alter method.__name__, so that obj.method
+            # will not be equal to getattr(obj, obj.method.__name__). We check
+            # for that case here and traverse to find the right method here.
+            for name in dir(obj):
+                meth = getattr(obj, name)
+                if inspect.ismethod(meth) and meth == callback:
+                    return obj, name
+            raise RuntimeError(
+                f"During bind method {callback} of object {obj} an error happen"
+            )
+        return obj, callback.__name__
+
+    @staticmethod
+    def _check_signature(fun: Callable) -> bool:
+        """
+        Check if function will accept event parameter
+        """
+        signature = inspect.signature(fun)
+        parameters_list = list(signature.parameters.values())
 
-    def _normalize_cb(self, callback) -> Union[CallbackRef, Callback]:
+        if sum(map(_is_pos_arg, parameters_list)) > 1:
```

two very minor suggestions (not a big deal since this only happens at connect):

first, we're looping through parameters_list twice here, and it could be reduced to once.
second, inasmuch as we're inspecting signatures, you could also raise an error on >1 pos arg.  If you want, you can borrow the [`_acceptable_posarg_range`](https://github.com/tlambert03/psygnal/blob/28472005a335769a477de04306c8ffc27ee184f9/psygnal/_signal.py#L820) function from psygnal which will loop once and give you the (min, max) number of posargs.  (you could remove the logic for max too if you want)

but... this is also fine with me

---

As Signature is often short (no more than 20 elements) I think that iterating over it once or twice has no matte for performance (when done on connection) but increase readability. But maybe @sofroniewn could told which approach is easier to understand?

---

that's fine, stick with it if you prefer it